### PR TITLE
katib on sqlflow design docs

### PR DESCRIPTION
**Pull request by samplise**

### CODE REVIEW

File: doc/design/katib_on_sqlflow.md [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    num_round=[20, 100],
+    max_depth=[],
+    validate_select="select * from test_table"
+LABEL class
+INTO my_xgb_model;
+```
+The above example,
+- This query tries to train a XGBoost model in Katib. 
+- `katib.xgboost.gtree`:
+    - `katib` indicates to create a Katib step via couler.
```
Comment (ywskycn) [inline doc/design/katib_on_sqlflow.md]: Not sure when using `katib` or `auto` here..

File: doc/design/katib_on_sqlflow.md [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    - `num_round` and `max_depth` are parameters to be optimized. If programmers specify the value range for a parameter (e.g., `num_round`), the range will be set for this parameter during training; otherwise, programmers assign an empty list to a parameter (e.g., `max_depth`), this parameter will be assigned default range during training. If programmers do not specify any parameters to be optimized, the system will optimize default parameters for this model.   
+    - `validate_select` is required in Katib jobs. This indicates the data used for testing in model training in Katib.  
+- `my_xgb_model` is the name of the trained model. 
+
+
+## Required update in SQLFlow couler codegen:
+
+In SQLFlow, couler codegen will generate python program from input SQL statements. In the generated python program, it invokes couler APIs (`couler.run_container(...)`) to create Argo steps. However, the current generated python program only calls `couler.run_container` to create a single container step, which does not work with job type steps (e.g., Katib). It needs to invoke different couler APIs according to input SQL queries.
```
Comment (ywskycn) [inline doc/design/katib_on_sqlflow.md]: Need to tell detailed difference between container and job here

File: doc/design/katib_on_sqlflow.md [hunk 3]
```diff
@@ -0,0 +1,60 @@
+# Katib on SQLFlow
+This design is about optimizing model hyperparameters in SQLFlow by creating Katib step via couler.
+
+## Requirements:
+
+Currently, SQLFlow allows programmers to use SQL queries to enable ML model training, prediction, and explanation. However, hyperparameters optimization is also a necessary function that can help users to tune hyperparamters of their models in an easy way. This function will be supported by Katib in SQLFlow.
+
+## Hyperparameter Optimization in Katib
+
+[Katib](https://github.com/kubeflow/katib) is a Kubernetes Native System for Hyperparameter Tuning and Neural Architecture Search. The system is inspired by Google Vizier and supports multiple ML/DL frameworks (e.g. TensorFlow, Apache MXNet, and PyTorch) and models (e.g., XGBoost).
```
Comment (ywskycn) [inline doc/design/katib_on_sqlflow.md]: Need to add a para to tell what a Katib job, the inputs, the outputs, and what the job can do.

File: doc/design/katib_on_sqlflow.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+In order to use Katib, programmers need to specify it is a Katib job by `TRAIN katib.{tf, pytorch, xgboost}.{...}` in SQL query. Here we use a simple example to explain the syntax of creating Katib job in SQLFlow.  
+
+``` sql
+SELECT * FROM train_table
+TRAIN katib.xgboost.gbtree
+WITH
+    objective=multi:softmax,
+    eta=1,
+    num_round=[20, 100],
+    max_depth=[],
```
Comment (ywskycn) [inline doc/design/katib_on_sqlflow.md:35]: Also we need a story for the default range, and users may not be required to add a range for each parameter.

File: doc/design/katib_on_sqlflow.md [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    - `validate_select` is required in Katib jobs. This indicates the data used for testing in model training in Katib.  
+- `my_xgb_model` is the name of the trained model. 
+
+
+## Required update in SQLFlow couler codegen:
+
+In SQLFlow, couler codegen will generate python program from input SQL statements. In the generated python program, it invokes couler APIs (`couler.run_container(...)`) to create Argo steps. However, the current generated python program only calls `couler.run_container` to create a single container step, which does not work with job type steps (e.g., Katib). It needs to invoke different couler APIs according to input SQL queries.
+
+In Katib case, couler codegen needs to check if the input SQL query is to create a Katib job. If it is, codegen needs to generate codes to invoke a couler katib API, like `couler.run_katib(...)`
```
Comment (ywskycn) [inline doc/design/katib_on_sqlflow.md]: So `run_katib` is a function defined in couler?

File: doc/design/katib_on_sqlflow.md [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+## Design for `couler.run_katib(...)`
+
+Considering Katib itself supports multiple models and frameworks, and more may come in the future, we use a unique API to create Katib steps in couler. The API is as following:
+
+`couler.run_katib("model_or_framework"=None, katib_params={}, model_or_framework_params={})`
+
+In this API, it includes three arguments:
+- `model_or_framework`: string, indicates the model (e.g., XGBoost) or frameworks (e.g., tf or pytorch) used in model training on Katib.
```
Comment (ywskycn) [inline doc/design/katib_on_sqlflow.md:68]: One way could be:
```xgboost.py

def train(....):
    # Submit a XGBoost training job based on the input parameters

def auto(...):
    # Submit a Katib job which tune the XGBoost models
```

In such a way, users can use `xgboost.train()` or `xgboost.auto()`...

File: doc/design/katib_on_sqlflow.md [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+SQLFlow does not create training jobs on Katib directly. Instead, SQLFlow creates Katib steps in a workflow via couler, and the workflow will be run by Argo on Kubernetes. More details can be found in [couler and SQLFlow](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/design/couler_sqlflow.md). 
+
+## SQLFlow Syntax
+
+In order to use Katib, programmers need to specify it is a Katib job by `TRAIN katib.{tf, pytorch, xgboost}.{...}` in SQL query. Here we use a simple example to explain the syntax of creating Katib job in SQLFlow.  
+
+``` sql
+SELECT * FROM train_table
+TRAIN katib.xgboost.gbtree
```
Comment (wangkuiyi) [inline doc/design/katib_on_sqlflow.md]: Please be aware that according to the [model zoo design](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/design/model_zoo.md#background), the model names a class in a Docker image and is in the form of `a_data_scientist/regressors:v0.2/MyDNNRegressor`.

Comment (wangkuiyi) [inline doc/design/katib_on_sqlflow.md]: Also, it is `TO TRAIN` instead of `TRAIN`.

File: doc/design/katib_on_sqlflow.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+WITH
+    objective=multi:softmax,
```
Comment (wangkuiyi) [inline doc/design/katib_on_sqlflow.md:32]: What does the prefix `multi:` mean?

File: doc/design/katib_on_sqlflow.md [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    eta=1,
+    num_round=[20, 100],
+    max_depth=[],
+    validate_select="select * from test_table"
```
Comment (wangkuiyi) [inline doc/design/katib_on_sqlflow.md]: What does `validate_select` mean? Do you mean `validation_dataset`?

File: doc/design/katib_on_sqlflow.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+    - `xgboost` indicates to train a XGBoost model.
+    - `gbtree` indicates the booster type in XGBoost model training. 
+- In the `WITH` clause:
+    - `objective` and `eta` are parameters in XGBoost. More details about XGBoost parameters see: [here](https://xgboost.readthedocs.io/en/latest/parameter.html#general-parameters). Those parameters are optional. If users do not specify those parameters, those parameters will be filled by some default value. 
+    - `num_round` and `max_depth` are parameters to be optimized. If programmers specify the value range for a parameter (e.g., `num_round`), the range will be set for this parameter during training; otherwise, programmers assign an empty list to a parameter (e.g., `max_depth`), this parameter will be assigned default range during training. If programmers do not specify any parameters to be optimized, the system will optimize default parameters for this model.   
+    - `validate_select` is required in Katib jobs. This indicates the data used for testing in model training in Katib.  
+- `my_xgb_model` is the name of the trained model. 
+
+
+## Required update in SQLFlow couler codegen:
```
Comment (wangkuiyi) [inline doc/design/katib_on_sqlflow.md]: couler => Couler

In quite many other places, we need the same renaming.

**Review by wangkuiyi — COMMENTED** [review]
Please following the model zoo design and rethink about the SQL syntax extension to use Katib.

File: doc/design/katib_on_sqlflow.md [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+In SQLFlow, couler codegen will generate python program from input SQL statements. In the generated python program, it invokes couler APIs (`couler.run_container(...)`) to create Argo steps. However, the current generated python program only calls `couler.run_container` to create a single container step, which does not work with job type steps (e.g., Katib). It needs to invoke different couler APIs according to input SQL queries.
+
+In Katib case, couler codegen needs to check if the input SQL query is to create a Katib job. If it is, codegen needs to generate codes to invoke a couler katib API, like `couler.run_katib(...)`
+
```
Comment (merlintang) [inline doc/design/katib_on_sqlflow.md:60]: I am confused about this part, since SQLflow query -> couler.run_container will not able to generate a couler.run_katib. However, the code gen logic of Sqlflow would able to generate a step like couler.run_katib on the run time. it means that pod of Argo would generate other workflow with couler.run_katib.

File: doc/design/katib_on_sqlflow.md [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+[Katib](https://github.com/kubeflow/katib) is a Kubernetes Native System for Hyperparameter Tuning and Neural Architecture Search. The system is inspired by Google Vizier and supports multiple ML/DL frameworks (e.g. TensorFlow, Apache MXNet, and PyTorch) and models (e.g., XGBoost).
+
+## Couler and SQLFlow
+
+SQLFlow does not create training jobs on Katib directly. Instead, SQLFlow creates Katib steps in a workflow via couler, and the workflow will be run by Argo on Kubernetes. More details can be found in [couler and SQLFlow](https://github.com/sql-machine-learning/sqlflow/blob/develop/doc/design/couler_sqlflow.md). 
+
+## SQLFlow Syntax
+
+In order to use Katib, programmers need to specify it is a Katib job by `TRAIN katib.{tf, pytorch, xgboost}.{...}` in SQL query. Here we use a simple example to explain the syntax of creating Katib job in SQLFlow.  
+
```
Comment (merlintang) [inline doc/design/katib_on_sqlflow.md:27]: creating Katib job for XGBoost training in SQLFlow

File: doc/design/katib_on_sqlflow.md (hunk 1 above)
Comment (samplise) [reply doc/design/katib_on_sqlflow.md]: Maybe auto is better than katib.

File: doc/design/katib_on_sqlflow.md (hunk 8 above)
Comment (samplise) [reply doc/design/katib_on_sqlflow.md:32]: `multi:softmax` is a value for the objective, which sets XGBoost to do multiclass classification using the softmax objective.

File: doc/design/katib_on_sqlflow.md (hunk 11 above)
Comment (samplise) [reply doc/design/katib_on_sqlflow.md:60]: We can use `run_container` to run a container and within this container to create a katib job. However, in this case, the katib job is not a step of Argo and not managed by Argo.

**Review by wangkuiyi — APPROVED** [review]
This is a good starting point. Let us merge it and follow up the comments in successive PRs.

File: doc/design/katib_on_sqlflow.md (hunk 1 above)
Comment (gaocegege) [reply doc/design/katib_on_sqlflow.md]: Prefer auto since katib may have a different name in the future.
